pkg/receive/multitsdb.go

```diff
@@ -623,6 +648,42 @@ func (t *MultiTSDB) TenantAppendable(tenantID string) (Appendable, error) {
 	return tenant.readyStorage(), nil
 }
 
+func (t *MultiTSDB) SetHashringConfig(cfg []HashringConfig) error {
+	t.hashringConfigs = cfg
+
+	// If a tenant's already existed in MultiTSDB, update its label set
+	// from the latest []HashringConfig.
+	// In case one tenant appears in multiple hashring configs,
+	// only the label set from the first hashring config is applied.
+	// This is the same logic as startTSDB.
+	updatedTenants := make([]string, 0)
+	for _, hc := range t.hashringConfigs {
+		for _, tenantId := range hc.Tenants {
+			if slices.Contains(updatedTenants, tenantId) {
+				continue
+			}
+			if t.tenants[tenantId] != nil {
+				updatedTenants = append(updatedTenants, tenantId)
+
+				lset := labelpb.ExtendSortedLabels(t.labels, labels.FromStrings(t.tenantLabelName, tenantId))
+				var err error
+				lset, err = parseAndExtendSortedLabels(lset, hc.ExternalLabels, t.logger)
+				if err != nil {
+					return errors.Wrap(err, "failed to parse external labels for tenant "+tenantId)
```

Comment:
This section is just a little confusing to me, maybe instead of reusing `lset` we could something like:
```suggestion
				initialLset := labelpb.ExtendSortedLabels(t.labels, labels.FromStrings(t.tenantLabelName, tenantId))
				lset, err := parseAndExtendSortedLabels(initialLset, hc.ExternalLabels, t.logger)
				if err != nil {
					return errors.Wrap(err, "failed to parse external labels for tenant "+tenantId)
```

Also both methods are named very similarly, just by looking at this I'm not sure how each method extends.

Reply:
Done

> Also both methods are named very similarly, just by looking at this I'm not sure how each method extends.

There are two differents:
- `labelpb.ExtendSortedLabels` extends labels `[]string`, `multiTSDB.extendLabels` extend labels `map[string]string` (I edited the function's name and logic, see Filip's comment [here](https://github.com/thanos-io/thanos/pull/5777#discussion_r1187389117))
- If there are duplicate labels in the initial label set and the external label set, `labelpb.ExtendSortedLabels` picks the external label, `multiTSDB.extendLabels` picks the initial label. This is why `labelpb.ExtendSortedLabels` is used to extend tenant ID label, and `multiTSDB.extendLabels` is used to extend tenants' external labels (to assure backward compatibility with receiver's labels).

So:
- Do you think we need other names?
- Should we move `multiTSDB.extendLabels` to a label-related pkg? 🤔

Reply:
Hm I think it might be fine to keep it here if it's specifically multi-TSDB related operation on labels. Maybe something like `extendWithTenantsExternalLabels`?

Reply:
> Hm I think it might be fine to keep it here if it's specifically multi-TSDB related operation on labels.

Yeah, we just introduced this function in this PR. So apparently only multiTSDB is using it now^^ But do you think some other components may need this (extend labels from map and pick the initial duplicate label) in the future? 

> Maybe something like `extendWithTenantsExternalLabels`?

If we keep it here then yeah, this is a better name. But if we move it to a label-related pkg to reuse later, then probably a more general name^^

Reply:
@matej-g TMWYT. We're just this 1 nit away from merging this PR^^

Reply:
Hey sorry for the delays, I think this is fine, given we have 2 approvals already and this PR has gone through a lot! It's not a major issue. Let's resolve this 👍.